### PR TITLE
Handle powers of units and expand macros better

### PR DESCRIPTION
**Pull request by FreddieAkeroyd**
Should now handle mm^2 etc.

See ISISComputingGroup/IBEX#2237

### CODE REVIEW

File: db_checker.py [hunk 1]
```diff
@@ -114,24 +116,20 @@ def allowed_unit(self, unit):
...
             if not (u in allowed_units):
-                # may be to the power of
-                if not (re.match(r"\^[-]?\d", u)):
-                    # may be prefixed
-                    if not (re.match(r'T|G|M|k|m|u|n|p|f', u[0]) and u[1:] in allowed_units):
-                        # special case of cm
-                        if not u == "cm":
-                            return False
+                # may be prefixed
+                if not (re.match(allowed_units_prefixes, u[0]) and u[1:] in allowed_units):
```
Comment (AdrianPotter) [inline db_checker.py]: Why are we using regex match to check for single character matches? Surely using a set like allowed units would be more consistent and Pythonic.

**Review by AdrianPotter — CHANGES_REQUESTED** [review]

File: db_checker.py [hunk 2]
```diff
@@ -20,10 +20,12 @@
 
 EGU_sub_list = {'longin', 'longout', 'ai', 'ao'}
 
-# list of the accepted units (standard prefixs to these units are also accepted and checked for below)
+# list of the accepted units. Standard prefixs to these units are also accepted and checked for below
+# but we need to allow 'cm' explicitly as itr is a non-standard unit prefix for metre
 allowed_units = {'A', 'angstrom', 'bar', 'bit', 'byte', 'C', 'count', 'degree', 'eV', 'frame', 'hour', 'Hz', 'inch',
-                 'interrupt', 'K', 'L', 'm', 'min', 'minute', 'ohm', 'Oersted', '%', 'photon', 'pixel', 'radian', 's',
-                 'torr', 'step', 'T', 'V'}
+                 'interrupt', 'K', 'L', 'm', 'min', 'minute', 'ohm', 'Oersted', '%', 'photon', 'pixel', 'radian', 's','torr', 'step', 'T', 'V', 'cm', 'Pa'}
```
Comment (ThomasLohnert) [inline db_checker.py]: Style: should have a line break after 's',

File: db_checker.py [hunk 3]
```diff
@@ -20,10 +20,12 @@
...
+
+allowed_units_prefixes = r'T|G|M|k|m|u|n|p|f';
```
Comment (ThomasLohnert) [inline db_checker.py]: Style: remove trailing semicolon

File: db_checker.py [hunk 4]
```diff
@@ -114,24 +116,20 @@ def allowed_unit(self, unit):
...
-        if "$" in unit:
-            return True
+        # split unit amalgamations
+        units = re.split(r'[/ ()]', u)
 
-        # otherwise split unit amalgamations
-        units = re.split(r'[/ ()]', unit)
         for u in units:
+            # remove any powers of units e.g. mm^2 -> mm
+            u = re.sub(r'\^[-]?\d', '', u)
```
Comment (ThomasLohnert) [inline db_checker.py]: This does not check whether the power symbol is actually in the right place and passes e.g. ^2MHz, M^2H^2z etc. Also allows endlessly stacking power units e.g. MHz^2^-3^2^3, which although technically a valid unit, I'm not sure whether it's a sensible thing to allow?

**Review by ThomasLohnert — CHANGES_REQUESTED** [review]
Happened to review this at the same time as Adrian - added some additional comments in the code.

File: db_checker.py (hunk 1 above)
Comment (FreddieAkeroyd) [reply db_checker.py]: A good point, given I refactored the string into a constant I should have gone further and done that

File: db_checker.py (hunk 4 above)
Comment (FreddieAkeroyd) [reply db_checker.py]: I can add a count to sub to make it do only one replace, then ^2^3 should fail. I think I can add a [A-Za-z]+ to the satisfy the other matter. In the future, maybe we should be using https://pypi.python.org/pypi/cf_units ?

File: db_checker.py (hunk 3 above)
Comment (FreddieAkeroyd) [reply db_checker.py]: You can tell I'm a C++ programmer at heart :-D

File: db_checker.py (hunk 4 above)
Comment (ThomasLohnert) [reply db_checker.py]: The changes cover most of these cases now, but still pass if there is just one power modifier somewhere in the middle of the unit (as long as there's at least one letter that comes before it). Adding an "end of string" character `$` to the end of the regex should do the trick though (assuming that the power modifier will always be the last part of the unit description)

Comment (FreddieAkeroyd) [reply db_checker.py]: Good idea, and good testing

Comment (FreddieAkeroyd) [reply db_checker.py]: I'll assume we don't need to handle fractional unit powers ( 1.5 or 3/2) for now!

**Review by ThomasLohnert — APPROVED** [review]

**Comment by FreddieAkeroyd** [comment]
@AdrianPotter is there anything still pending on this?

**Comment by DominicOram** [comment]
Minor issues: 
* Now that cm is a valid unit all the prefixes are also valid e.g. "mcm", "kcm", "Tcm" etc.
* In some cases brackets cause a string index out of range exception to be thrown. e.g. "K/(m s)"  
* More of a question, should we allow units such as "1/m"? The program currently does not

It would also be good if we could add unit tests to the checker, which I know might be tricky as it is itself a test suite but it will help avoid this issue coming up again?

**Comment by FreddieAkeroyd** [comment]
I wonder if we dhould just use something like https://pypi.python.org/pypi/cf_units or https://pypi.python.org/pypi/cfunits/1.0.1

**Review by AdrianPotter — APPROVED** [review]

**Comment by AdrianPotter** [comment]
- I've set up a `db` file to specifically test units. This can be run using `test_checker.bat`
- I've improved and simplified the unit testing logic. It also caught several failing cases in the new tests. This includes Dom's points
- I haven't allowed 1/m for the time being
- I looked at using one of the libraries Freddie suggested but I had trouble getting them using `pip install` so decided they were more trouble than they were worth.

**Pull request merged**